broker/src/main/java/io/zeebe/broker/system/partitions/impl/StateControllerImpl.java

```diff
@@ -0,0 +1,312 @@
+   *
+   * @param snapshotChunk the chunk to consume
+   */
+  private void consumeSnapshotChunk(final SnapshotChunk snapshotChunk) {
+    final String snapshotId = snapshotChunk.getSnapshotId();
+    final String chunkName = snapshotChunk.getChunkName();
+
+    final ReplicationContext context =
+        receivedSnapshots.computeIfAbsent(
+            snapshotId,
+            id -> {
+              final var startTimestamp = System.currentTimeMillis();
+              final ReceivedSnapshot transientSnapshot =
+                  store.newReceivedSnapshot(snapshotChunk.getSnapshotId());
+              return newReplication(startTimestamp, transientSnapshot);
+            });
+    if (context == INVALID_SNAPSHOT) {
+      LOG.trace(
+          "Ignore snapshot chunk {}, because snapshot {} is marked as invalid.",
+          chunkName,
+          snapshotId);
+      return;
+    }
+
+    final var receivedSnapshot = context.getReceivedSnapshot();
+    try {
+      if (receivedSnapshot.apply(snapshotChunk)) {
+        validateWhenReceivedAllChunks(snapshotChunk, context);
+      } else {
+        markSnapshotAsInvalid(context, snapshotChunk);
+      }
+    } catch (final IOException e) {
+      LOG.error("Unexepected error on writing the received snapshot chunk {}", snapshotChunk, e);
+      markSnapshotAsInvalid(context, snapshotChunk);
+    }
+  }
+
+  private void markSnapshotAsInvalid(
+      final ReplicationContext replicationContext, final SnapshotChunk chunk) {
+    replicationContext.getReceivedSnapshot().abort();
+    receivedSnapshots.put(chunk.getSnapshotId(), INVALID_SNAPSHOT);
+    metrics.decrementCount();
+  }
+
+  private void validateWhenReceivedAllChunks(
+      final SnapshotChunk snapshotChunk, final ReplicationContext context) {
+    final int totalChunkCount = snapshotChunk.getTotalCount();
+
+    if (context.incrementCount() == totalChunkCount) {
+      LOG.debug(
+          "Received all snapshot chunks ({}/{}), snapshot is valid",
+          context.getChunkCount(),
+          totalChunkCount);
+      if (!tryToMarkSnapshotAsValid(snapshotChunk, context)) {
+        LOG.debug("Failed to mark snapshot {} as valid", snapshotChunk.getSnapshotId());
+      }
+    } else {
+      LOG.debug(
+          "Waiting for more snapshot chunks, currently have {}/{}",
+          context.getChunkCount(),
+          totalChunkCount);
+    }
+  }
+
+  private boolean tryToMarkSnapshotAsValid(
+      final SnapshotChunk snapshotChunk, final ReplicationContext context) {
+    try {
+      context.getReceivedSnapshot().persist();
+    } catch (final Exception exception) {
+      markSnapshotAsInvalid(context, snapshotChunk);
+      LOG.warn("Unexpected error on persisting received snapshot.", exception);
+      return false;
+    } finally {
+      final var elapsed = System.currentTimeMillis() - context.getStartTimestamp();
+      receivedSnapshots.remove(snapshotChunk.getSnapshotId());
+      metrics.decrementCount();
```

Comment:
In case of an exception, the decrement already happened as part of `markSnapshotAsInvalid(...)` In general I am not sure whether the increments and decrements in this class will always balance out perfectly. Maybe they do now, but it is fragile.

I think it would be better to add a `destroy()/deactivate()` method to the `ReplicationContext` which counts the metric down when it is first called, and does nothing when it is called multiple times.

Reply:
:+1: yes I also didn't liked this approach